simulator/func_sim/alu.h

```diff
@@ -16,6 +16,93 @@
 template<size_t N, typename T>
 T align_up(T value) { return ((value + ((1ull << N) - 1)) >> N) << N; }
 
+enum shuffle_consts { MASK_L_0 = 0x00ff0000, MASK_R_0 = 0x0000ff00,
+                      MASK_L_1 = 0x0f000f00, MASK_R_1 = 0x00f000f0,
+                      MASK_L_2 = 0x30303030, MASK_R_2 = 0x0c0c0c0c,
+                      MASK_L_3 = 0x44444444, MASK_R_3 = 0x22222222,
+};
+
+static uint32 shuffle_stage( uint32 src, size_t index)
+{
+    uint32 maskL, maskR;
+    switch(index)
+    {
+        case 0:
+            maskL = MASK_L_0;
+            maskR = MASK_R_0;
+            break;
+        case 1:
+            maskL = MASK_L_1;
+            maskR = MASK_R_1;
+            break;
+        case 2:
+            maskL = MASK_L_2;
+            maskR = MASK_R_2;
+            break;
+        case 3:
+            maskL = MASK_L_3;
+            maskR = MASK_R_3;
+            break;
+        default:
+            throw std::runtime_error( "Unexpexted index in 'shuffle_stage' function");
+            break;
+    }
+    index = 1 << (3 - index);
+    uint32 x = src & ~( maskL | maskR);
+    x |= ( ( src << index) & maskL) | ( ( src >> index) & maskR);
+    return x;
+}
+
+static uint32 bit_shuffle( uint32 src1, uint32 src2)
+{
+    uint32 pre_result = src1;
+    uint32 shamt = src2 & 15;
+    for( size_t i = 0; i < 4 ; ++i)
+    {
+        if ( shamt & ( 1 << ( 3 - i)))
+            pre_result = shuffle_stage( pre_result, i);
+    }
+    return pre_result;
+}
+
+static uint64 bit_shuffle( uint64& src1, uint64 src2)
+{
+    uint64 result = src1;
+    uint32 shamt = src2 & 31;
+    uint32 left_part = result >> 32;
+    uint32 right_part = result;
+    if( shamt & (1 << 4))
+    {
+        uint32 buf = left_part & bitmask<uint32>(16);
+        left_part = (left_part & ~(bitmask<uint32>(16))) | ((right_part >> 16) & bitmask<uint32>(16));
+        right_part = (right_part & (bitmask<uint32>(16))) | (buf << 16);
+    }
+    left_part = bit_shuffle(left_part, shamt & 15);
+    right_part = bit_shuffle(right_part, shamt & 15);
+    result = left_part;
+    result <<= 32;
+    return result + right_part;
+}
+
+__attribute__( ( unused)) static uint128 bit_shuffle( uint128 src1, uint128 src2)
```

Comment:
I'm afraid VS do not know about that syntax. There should be something in C++17 probably.
Besides that, I'm against checking in unused/untested code. If we want to change it, there will be no warranty we did not break it.